File: spring-mvc-plain-java-example/src/main/java/de/cfranzen/examples/webflux/CustomerController.java
package de.cfranzen.examples.webflux;

import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

@RestController
class CustomerController {

    private final CustomerRepository repository;

    CustomerController(final CustomerRepository repository) {
        this.repository = repository;
    }

    @PostMapping("/customer")
    Customer createNewCustomer(@RequestBody CustomerCreationRequest request) {
        return repository.save(new Customer(request.firstName(), request.lastName()));
    }
}

record CustomerCreationRequest(
        String firstName,
        String lastName
) {
}

File: spring-webflux-kotlin-example/src/main/kotlin/de/cfranzen/examples/webflux/kotlin/CustomerController.kt
package de.cfranzen.examples.webflux.kotlin

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.single
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
internal class CustomerController(
    private val repository: CustomerRepository
) {

    @PostMapping("/customer")
    suspend fun createNewCustomer(@RequestBody request: Flow<CustomerCreationRequest>): Customer {
        val dto = request.single()
        return repository.save(Customer(dto.firstName, dto.lastName))
    }
}

internal data class CustomerCreationRequest(
    val firstName: String,
    val lastName: String
)
